src/number_theory/modular.lean

```diff
@@ -0,0 +1,380 @@
+  `analysis.complex.upper_half_plane`). We then define the standard fundamental domain
+  (`modular_group.fundamental_domain`, `𝒟`) for this action and show
+  (`modular_group.exists_smul_mem_fundamental_domain`) that any point in `ℍ` can be
+  moved inside `𝒟`.
+
+  Standard proofs make use of the identity
+
+  `g•z = a/c - 1/(c(cz+d))`
+
+  for `g=[[a,b],[c,d]]` in `SL(2)`, but this requires separate handling of whether `c = 0`.
+  Instead, our proof makes use of the following perhaps novel identity (see
+  `modular_group.smul_eq_acbd_add`):
+
+  `g • z = (a c + b d) / (c^2 + d^2) + (d z - c) / ((c^2 + d^2) (c z + d))`
+
+  where there is no issue of division by zero.
+
+  Another feature is that we delay until the very end the consideration of special matrices
+  `T=[[1,1],[0,1]]` (see `modular_group.T`) and `S=[[0,-1],[1,0]]` (see `modular_group.S`), by
+  instead using abstract theory on the properness of certain maps (phrased in terms of the filters
+  `filter.cocompact`, `filter.cofinite`, etc) to deduce existence theorems, first to prove the
+  existence of `g` maximizing `(g•z).im` (see `modular_group.exists_g_with_max_im`), and then among
+  those, to minimize `|(g•z).re|` (see `modular_group.exists_g_with_given_cd_and_min_re`).
+-/
+
+open complex matrix matrix.special_linear_group upper_half_plane
+noncomputable theory
+
+local notation `|` x `|` := _root_.abs x
+local notation `SL(` n `, ` R `)`:= special_linear_group (fin n) R
+
+open_locale upper_half_plane
+
+local attribute [instance] fintype.card_fin_even
+
+namespace modular_group
+
+section upper_half_plane_action
+
+/-- The action of `SL(2, ℤ)` on the upper half-plane, as a restriction of the `SL(2, ℝ)`-action. -/
+instance : mul_action SL(2, ℤ) ℍ :=
+mul_action.comp_hom ℍ (map (int.cast_ring_hom ℝ))
+
+@[simp] lemma coe_smul (g : SL(2, ℤ)) (z : ℍ) : ↑(g • z) = num g z / denom g z := rfl
+@[simp] lemma re_smul (g : SL(2, ℤ)) (z : ℍ) : (g • z).re = (num g z / denom g z).re := rfl
+@[simp] lemma smul_coe (g : SL(2, ℤ)) (z : ℍ) : (g : SL(2,ℝ)) • z = g • z := rfl
+
+@[simp] lemma neg_smul (g : SL(2, ℤ)) (z : ℍ) : -g • z = g • z :=
+show ↑(-g) • _ = _, by simp [neg_smul g z]
+
+lemma im_smul (g : SL(2, ℤ)) (z : ℍ) : (g • z).im = (num g z / denom g z).im := rfl
+
+lemma im_smul_eq_div_norm_sq (g : SL(2, ℤ)) (z : ℍ) :
+  (g • z).im = z.im / (complex.norm_sq (denom g z)) :=
+im_smul_eq_div_norm_sq g z
+
+end upper_half_plane_action
+
+section bottom_row
+
+/-- The `bottom_row` of `g=[[*,*],[c,d]]` in `SL(2,ℤ)` is the `coprime_pair` `(c,d)`. -/
+def bottom_row (g : SL(2, ℤ)) : coprime_pair :=
+coprime_pair.mk
+  (@coe _ (matrix (fin 2) (fin 2) ℤ) _ g 1 0, @coe _ (matrix (fin 2) (fin 2) ℤ) _ g 1 1)
+  begin
+    use [- g 0 1, g 0 0],
+    have := det_fin_two g,
+    have := g.det_coe,
+    simp only [coe_fn_eq_coe] at *,
+    linarith
+  end
+
+lemma bottom_row_surj : function.surjective bottom_row :=
```

Comment:
As an alternative to `coprime_pair`, have you considered using `set.surj_on` to exclude the non-coprime elements of the codomain?

Reply:
Just pushed a revision with a relative version of `filter.tendsto.exists_forall_le`, which allows this to go through.